src/i2c_bit_bang.cpp

```diff
@@ -0,0 +1,255 @@
+#include <chrono>
+
+#include <libhal-soft/i2c_bit_bang.hpp>
+#include <libhal-util/bit.hpp>
+#include <libhal-util/i2c.hpp>
+#include <libhal-util/steady_clock.hpp>
+#include <libhal/error.hpp>
+#include <libhal/units.hpp>
+
+namespace hal {
+// public
+i2c_bit_bang::i2c_bit_bang(i2c_pins p_pins,
+                           steady_clock& p_clock,
+                           bus_info p_bus)
+  : m_scl(p_pins.scl)
+  , m_sda(p_pins.sda)
+  , m_clock(&p_clock)
+  , m_bus(p_bus)
+{
+}
+
+void i2c_bit_bang::driver_configure(const settings& p_settings)
+{
+  using namespace std::chrono_literals;
+
+  if (p_settings.clock_rate > m_clock->frequency()) {
+    throw hal::operation_not_supported(this);
+  }
+
+  // calculate period in nanosecond
+  using period = std::chrono::nanoseconds::period;
+  auto period_ns = hal::wavelength<period>(p_settings.clock_rate);
+  auto scl_high_time = period_ns * m_bus.duty_cycle;
+  auto scl_low_time = period_ns - scl_high_time;
+
+  // calculate ticks for high and low
+  const auto frequency = m_clock->frequency();
+  const auto tick_period = hal::wavelength<period>(frequency);
+
+  m_scl_high_ticks = static_cast<uint64_t>(scl_high_time / tick_period);
+
+  m_scl_low_ticks = static_cast<uint64_t>(scl_low_time / tick_period);
+
+  // calculate the delay due to the uptime function call
+  const auto callibration_start_tick = m_clock->uptime();
+  const auto callibration_end_tick = m_clock->uptime();
+  m_uptime_ticks = callibration_end_tick - callibration_start_tick;
+
+  // calculating output_pin going to true delay time
+  const auto before_output_high = m_clock->uptime();
+  m_scl->level(true);
+  const auto after_output_high = m_clock->uptime();
+  const auto output_true_ticks = after_output_high - before_output_high;
+
+  m_scl_high_ticks -= output_true_ticks;
+
+  // calculating output_pin going to false delay time
+  const auto before_output_low = m_clock->uptime();
+  m_scl->level(false);
```

Comment:
```suggestion
  const auto before_output_high = m_clock->uptime();
  m_scl->level(true);
  const auto after_output_high = m_clock->uptime();
  const auto output_true_ticks = after_output_high - before_output_high;

  m_scl_high_ticks -= output_true_ticks;

  // calculating output_pin going to false delay time
  const auto before_output_low = m_clock->uptime();
  m_scl->level(true);
```

Do not actually clock the clock line. This could potentially confuse the devices on the bus. Its best to do an operation that does nothing but at least gives us the info of how long it takes to set the level.

Reply:
Gotcha, I figured that the function may take different times for low and high and wanted to record both but this is probably not the case.